### ISSUE TITLE
[Phase 4] Threshold Optimization

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/fraud-detection-adaptive#6]
## Summary

Implement adaptive threshold selection based on cost-weighted precision-recall analysis. This phase creates the decision layer that converts continuous scores into optimal binary decisions.

**Parent Issue**: #1

## Objectives

- Implement cost-weighted F-score calculation
- Create optimal threshold finder using P-R curves
- Build unified scorer combining unsupervised and supervised models
- Enable threshold updates based on new validation data

## Dependencies

- Phase 2 must be complete (unsupervised ensemble)
- Phase 3 must be complete (supervised XGBoost)

## Task Checklist

- [ ] Create `src/scoring/__init__.py` - package init
- [ ] Implement `src/scoring/threshold_optimizer.py` - P-R curve analysis
- [ ] Implement `src/scoring/cost_calculator.py` - business cost computation
- [ ] Implement `src/scoring/scorer.py` - unified scoring interface
- [ ] Create `tests/test_threshold.py` - unit tests

### Threshold Optimizer

📄 **File:** `src/scoring/threshold_optimizer.py`

```python
"""Threshold optimization for fraud detection scoring."""

from sklearn.metrics import precision_recall_curve
import numpy as np
from typing import Tuple, Optional
from dataclasses import dataclass


@dataclass
class ThresholdResult:
    """Result of threshold optimization."""
    optimal_threshold: float
    expected_cost: float
    precision_at_threshold: float
    recall_at_threshold: float


class ThresholdOptimizer:
    """
    Find optimal decision threshold using cost-weighted metrics.

    Instead of maximizing F1 (equal weight to precision/recall),
    we minimize expected cost based on business impact.
    """

    def __init__(
        self,
        fn_cost: float = 100.0,
        fp_cost: float = 1.0,
        fraud_prior: float = 0.0001  # Expected fraud rate
    ):
        """
        Args:
            fn_cost: Cost per false negative (missed fraud)
            fp_cost: Cost per false positive (blocked legitimate)
            fraud_prior: Prior probability of fraud (for cost calculation)
        """
        self.fn_cost = fn_cost
        self.fp_cost = fp_cost
        self.fraud_prior = fraud_prior

        self.optimal_threshold: float = 0.5
        self._is_fitted = False

    def find_optimal_threshold(
        self,
        y_true: np.ndarray,
        y_scores: np.ndarray
    ) -> ThresholdResult:
        """
        Find threshold that minimizes expected cost.

        Uses precision-recall curve to evaluate all possible thresholds.
        Cost at each threshold:
            Cost = FN_rate * P(fraud) * FN_cost + FP_rate * P(legit) * FP_cost
                 = (1 - recall) * prior * fn_cost + (1 - precision) * (1-prior) * fp_cost

        Args:
            y_true: True binary labels
            y_scores: Predicted probabilities

        Returns:
            ThresholdResult with optimal threshold and metrics
        """
        precisions, recalls, thresholds = precision_recall_curve(y_true, y_scores)

        # Calculate cost at each threshold
        # Note: precision_recall_curve returns n+1 precisions/recalls but n thresholds
        # Last precision/recall is for threshold = 1 (everything predicted negative)

        costs = []
        for i, threshold in enumerate(thresholds):
            precision = precisions[i]
            recall = recalls[i]

            # FN rate = 1 - recall (among actual positives)
            # FP rate = (1 - precision) * (# predicted positive) / (# actual negative)
            # Simplified: we use precision and recall directly

            fn_rate = 1 - recall
            fp_rate = (1 - precision) if precision > 0 else 1.0

            cost = (
                fn_rate * self.fraud_prior * self.fn_cost +
                fp_rate * (1 - self.fraud_prior) * self.fp_cost
            )
            costs.append(cost)

        costs = np.array(costs)

        # Find minimum cost threshold
        best_idx = np.argmin(costs)
        self.optimal_threshold = thresholds[best_idx]
        self._is_fitted = True

        return ThresholdResult(
            optimal_threshold=self.optimal_threshold,
            expected_cost=costs[best_idx],
            precision_at_threshold=precisions[best_idx],
            recall_at_threshold=recalls[best_idx]
        )

    def get_threshold(self) -> float:
        """Return the current optimal threshold."""
        return self.optimal_threshold

    def evaluate_threshold(
        self,
        y_true: np.ndarray,
        y_scores: np.ndarray,
        threshold: Optional[float] = None
    ) -> dict:
        """Evaluate a specific threshold."""
        if threshold is None:
            threshold = self.optimal_threshold

        y_pred = (y_scores >= threshold).astype(int)

        tp = ((y_true == 1) & (y_pred == 1)).sum()
        fp = ((y_true == 0) & (y_pred == 1)).sum()
        tn = ((y_true == 0) & (y_pred == 0)).sum()
        fn = ((y_true == 1) & (y_pred == 0)).sum()

        precision = tp / (tp + fp) if (tp + fp) > 0 else 0.0
        recall = tp / (tp + fn) if (tp + fn) > 0 else 0.0
        total_cost = fn * self.fn_cost + fp * self.fp_cost

        return {
            'threshold': threshold,
            'precision': precision,
            'recall': recall,
            'tp': int(tp),
            'fp': int(fp),
            'tn': int(tn),
            'fn': int(fn),
            'total_cost': float(total_cost)
        }
```

### Unified Scorer

📄 **File:** `src/scoring/scorer.py`

```python
"""Unified fraud scoring interface."""

from typing import Optional, Dict, Any
from dataclasses import dataclass
import numpy as np
import json
from pathlib import Path

from src.models.unsupervised_ensemble import UnsupervisedEnsemble
from src.models.xgboost_classifier import CostSensitiveXGBoost
from src.scoring.threshold_optimizer import ThresholdOptimizer


@dataclass
class ScoringResult:
    """Result of scoring a transaction."""
    transaction_id: str
    fraud_score: float
    decision: str  # "approve", "review", "decline"
    unsupervised_score: float
    supervised_score: float
    threshold_used: float


class FraudScorer:
    """
    Unified fraud scoring combining unsupervised and supervised models.

    Scoring pipeline:
    1. Get unsupervised anomaly score
    2. Get supervised fraud probability
    3. Combine scores (weighted average or max)
    4. Apply threshold to make decision
    """

    def __init__(
        self,
        unsupervised_model: Optional[UnsupervisedEnsemble] = None,
        supervised_model: Optional[CostSensitiveXGBoost] = None,
        threshold_optimizer: Optional[ThresholdOptimizer] = None,
        unsupervised_weight: float = 0.3,
        supervised_weight: float = 0.7,
        review_threshold: Optional[float] = None  # Threshold for "review" bucket
    ):
        """
        Args:
            unsupervised_model: Fitted unsupervised ensemble
            supervised_model: Fitted XGBoost model
            threshold_optimizer: Fitted threshold optimizer
            unsupervised_weight: Weight for unsupervised score
            supervised_weight: Weight for supervised score
            review_threshold: Threshold between approve and review (if None, no review bucket)
        """
        self.unsupervised_model = unsupervised_model
        self.supervised_model = supervised_model
        self.threshold_optimizer = threshold_optimizer or ThresholdOptimizer()

        self.unsupervised_weight = unsupervised_weight
        self.supervised_weight = supervised_weight
        self.review_threshold = review_threshold

        self.decline_threshold = self.threshold_optimizer.get_threshold()

    def score(self, X: np.ndarray) -> np.ndarray:
        """
        Compute combined fraud scores for feature matrix.

        Returns scores in [0, 1] range, higher = more likely fraud.
        """
        scores = np.zeros(len(X))

        if self.unsupervised_model is not None:
            unsup_scores = self.unsupervised_model.score(X)
            scores += self.unsupervised_weight * unsup_scores

        if self.supervised_model is not None:
            sup_scores = self.supervised_model.score(X)
            scores += self.supervised_weight * sup_scores

        return scores

    def predict(
        self,
        X: np.ndarray,
        transaction_ids: Optional[list] = None
    ) -> list[ScoringResult]:
        """
        Make predictions with full result details.

        Returns list of ScoringResult objects.
        """
        combined_scores = self.score(X)

        # Get individual model scores
        unsup_scores = (
            self.unsupervised_model.score(X)
            if self.unsupervised_model else np.zeros(len(X))
        )
        sup_scores = (
            self.supervised_model.score(X)
            if self.supervised_model else np.zeros(len(X))
        )

        results = []
        for i in range(len(X)):
            txn_id = transaction_ids[i] if transaction_ids else f"txn_{i}"

            # Determine decision
            if combined_scores[i] >= self.decline_threshold:
                decision = "decline"
            elif self.review_threshold and combined_scores[i] >= self.review_threshold:
                decision = "review"
            else:
                decision = "approve"

            results.append(ScoringResult(
                transaction_id=txn_id,
                fraud_score=float(combined_scores[i]),
                decision=decision,
                unsupervised_score=float(unsup_scores[i]),
                supervised_score=float(sup_scores[i]),
                threshold_used=float(self.decline_threshold)
            ))

        return results

    def update_threshold(self, y_true: np.ndarray, y_scores: np.ndarray):
        """Update decision threshold based on new validation data."""
        result = self.threshold_optimizer.find_optimal_threshold(y_true, y_scores)
        self.decline_threshold = result.optimal_threshold
        return result

    def save(self, path: str):
        """Save scorer configuration (models saved separately)."""
        config = {
            'unsupervised_weight': self.unsupervised_weight,
            'supervised_weight': self.supervised_weight,
            'decline_threshold': self.decline_threshold,
            'review_threshold': self.review_threshold
        }
        Path(path).write_text(json.dumps(config, indent=2))

    @classmethod
    def load_champion(cls, model_dir: str = "models/champion") -> 'FraudScorer':
        """Load the champion (production) model."""
        # Implementation depends on MLflow setup in Phase 6
        raise NotImplementedError("Implement after Phase 6")
```

## Verification Commands

```bash
# Run threshold tests
pytest tests/test_threshold.py -v

# Test threshold optimization
python -c "
import numpy as np
from src.scoring.threshold_optimizer import ThresholdOptimizer

# Synthetic scores
np.random.seed(42)
n = 10000
y_true = np.random.binomial(1, 0.001, n)
y_scores = np.where(y_true == 1, np.random.beta(5, 2, n), np.random.beta(2, 5, n))

optimizer = ThresholdOptimizer(fn_cost=100, fp_cost=1)
result = optimizer.find_optimal_threshold(y_true, y_scores)

print(f'Optimal threshold: {result.optimal_threshold:.3f}')
print(f'Expected cost: {result.expected_cost:.4f}')
print(f'Precision: {result.precision_at_threshold:.3f}')
print(f'Recall: {result.recall_at_threshold:.3f}')
"
```

## Definition of Done

- [ ] Threshold minimizes expected cost on validation set
- [ ] Threshold selection is reproducible with same inputs
- [ ] API returns both score and binary decision
- [ ] Three-way decision (approve/review/decline) works correctly
- [ ] Threshold can be updated without refitting models
- [ ] All tests pass

## Files to Modify

| File | Lines | Action | Description |
|------|-------|--------|-------------|
| `src/scoring/__init__.py` | - | Create | Package init |
| `src/scoring/threshold_optimizer.py` | - | Create | P-R curve optimizer |
| `src/scoring/scorer.py` | - | Create | Unified scorer |
| `tests/test_threshold.py` | - | Create | Unit tests |

## Concepts to Understand

Before implementing, review:
- **Precision-Recall curve**: Trade-off as threshold varies
- **Cost-sensitive learning**: Weighting errors by business impact
- **Decision boundaries**: Converting probability to action
